File: android/src/main/java/com/kiwi/fluttercrashlytics/FlutterCrashlyticsPlugin.java
package com.kiwi.fluttercrashlytics;

import android.content.Context;
import android.content.Intent;
import android.util.Log;
import com.crashlytics.android.Crashlytics;
import com.crashlytics.android.core.CrashlyticsCore;
import io.fabric.sdk.android.Fabric;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.PluginRegistry;

import java.util.List;
import java.util.Map;

public class FlutterCrashlyticsPlugin implements MethodChannel.MethodCallHandler {

    private final Context context;

    public FlutterCrashlyticsPlugin(Context context) {
        this.context = context;
    }

    public static void registerWith(PluginRegistry.Registrar registrar) {
        final MethodChannel channel = new MethodChannel(registrar.messenger(), "flutter_crashlytics");
        channel.setMethodCallHandler(new FlutterCrashlyticsPlugin(registrar.context()));
    }

    @Override
    public void onMethodCall(MethodCall methodCall, MethodChannel.Result result) {
        if (methodCall.method.equals("initialize")) {
            Fabric.with(context, new Crashlytics());
            result.success(null);
        } else {
            if (Fabric.isInitialized()) {
                onInitialisedMethodCall(methodCall, result);
            } else {
                result.success(null);
            }
        }
    }

    void onInitialisedMethodCall(MethodCall call, MethodChannel.Result result) {
        final CrashlyticsCore core = Crashlytics.getInstance().core;
        switch (call.method) {
            case "log":
                if (call.arguments instanceof String) {
                    core.log(call.arguments.toString());
                } else {
                    try {
                        final List<Object> info = (List<Object>) call.arguments;
                        core.log(parseStringOrEmpty(info.get(0)) + ": " + info.get(1) + " " + info.get(2));
                    } catch (ClassCastException ex) {
                        Log.d("FlutterCrashlytics", "" + ex.getMessage());
                    }
                }
                result.success(null);
                break;
            case "setInfo":
                final Object currentInfo = call.argument("value");
                if (currentInfo instanceof String) {
                    core.setString((String) call.argument("key"), (String) call.argument("value"));
                } else if (currentInfo instanceof Integer) {
                    core.setInt((String) call.argument("key"), (Integer) call.argument("value"));
                } else if (currentInfo instanceof Double) {
                    core.setDouble((String) call.argument("key"), (Double) call.argument("value"));
                } else if (currentInfo instanceof Boolean) {
                    core.setBool((String) call.argument("key"), (Boolean) call.argument("value"));
                } else if (currentInfo instanceof Float) {
                    core.setFloat((String) call.argument("key"), (Float) call.argument("value"));
                } else if (currentInfo instanceof Long) {
                    core.setLong((String) call.argument("key"), (Long) call.argument("value"));
                } else {
                    core.log("ignoring unknown type with key " + call.argument("key") + "and value "
                            + call.argument("value"));
                }
                result.success(null);
                break;
            case "setUserInfo":
                core.setUserEmail((String) call.argument("email"));
                core.setUserName((String) call.argument("name"));
                core.setUserIdentifier((String) call.argument("id"));
                result.success(null);
                break;
            case "reportCrash":
                final Map<String, Object> exception = (Map<String, Object>) call.arguments;
                final boolean forceCrash = tryParseForceCrash(exception.get("forceCrash"));
                final FlutterException throwable = Utils.create(exception);
                if (forceCrash) {
                    //Start a new activity to not crash directly under onMethod call, or it will crash JNI instead of a clean exception
                    final Intent intent = new Intent(context, CrashActivity.class);
                    intent.putExtra("exception", throwable);
                    intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);

                    context.startActivity(intent);
                } else {
                    Log.d("Crashlytics", "this thing is reporting");
                    core.logException(throwable);
                }

                result.success(null);
                break;
            default:
                result.notImplemented();
                break;
        }
    }

    String parseStringOrEmpty(Object obj) {
        if (obj != null) {
            return obj.toString();
        }
        return "";
    }

    boolean tryParseForceCrash(Object object) {
        if (object instanceof Boolean) {
            return (boolean) object;
        }
        return false;
    }

}

File: android/src/main/kotlin/com/kiwi/fluttercrashlytics/FlutterCrashlyticsPlugin.kt
package com.kiwi.fluttercrashlytics

import android.app.Activity
import android.content.Intent
import com.crashlytics.android.Crashlytics
import io.fabric.sdk.android.Fabric
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry.Registrar


class FlutterCrashlyticsPlugin(private val context: Activity) : MethodCallHandler {
    companion object {
        @JvmStatic
        fun registerWith(registrar: Registrar) {
            val channel = MethodChannel(registrar.messenger(), "flutter_crashlytics")
            channel.setMethodCallHandler(FlutterCrashlyticsPlugin(registrar.activity()))
        }
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "initialise" -> {
                Fabric.with(context, Crashlytics())

                result.success(null)
            }
            else -> {
                if (Fabric.isInitialized()) {
                    onInitialisedMethodCall(call, result)
                } else {
                    // Should not result in an error. Otherwise Opt Out clients would need to handle errors
                    result.success(null)
                }
            }
        }
    }

    private fun onInitialisedMethodCall(call: MethodCall, result: Result) {
        val core = Crashlytics.getInstance().core
        when {
            call.method == "log" -> {
                if (call.arguments is String) {
                    core.log(call.arguments as String)
                } else {
                    val info = call.arguments as List<Any>
                    core.log(info[0] as Int, info[1] as String, info[2] as String)
                }
                result.success(null)
            }
            call.method == "setInfo" -> {
                val info = call.arguments as Map<String, Any>
                when (info["value"]) {
                    is String ->
                        core.setString(info["key"] as String, info["value"] as String)
                    is Int ->
                        core.setInt(info["key"] as String, info["value"] as Int)
                    is Double ->
                        core.setDouble(info["key"] as String, info["value"] as Double)
                    is Boolean ->
                        core.setBool(info["key"] as String, info["value"] as Boolean)
                    is Float ->
                        core.setFloat(info["key"] as String, info["value"] as Float)
                    is Long ->
                        core.setLong(info["key"] as String, info["value"] as Long)
                    else -> core.log("ignoring unknown type with key ${info["key"]} and value ${info["value"]}")
                }
                result.success(null)
            }
            call.method == "setUserInfo" -> {
                val info = call.arguments as Map<String, String>
                core.setUserEmail(info["email"])
                core.setUserName(info["name"])
                core.setUserIdentifier(info["id"])
                result.success(null)
            }
            call.method == "reportCrash" -> {
                val exception = (call.arguments as Map<String, Any>)
                val forceCrash = exception["forceCrash"] as? Boolean ?: false

                val throwable = Utils.create(exception)

                if (forceCrash) {
                    //Start a new activity to not crash directly under onMethod call, or it will crash JNI instead of a clean exception
                    val intent = Intent(context, CrashActivity::class.java).apply {
                        putExtra("exception", throwable)
                    }

                    context.startActivityForResult(intent, -1)
                } else {
                    core.logException(throwable)
                }
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }
}
